Sources/Sentry/SentrySerialization.m

```diff
@@ -310,6 +292,153 @@ + (SentryLevel)levelFromData:(NSData *)eventEnvelopeItemData
     return sentryLevelForString(eventDictionary[@"level"]);
 }
 
++ (BOOL)writeObject:(id)jsonObject toStream:(NSOutputStream *)stream
+{
+    NSInteger writeResult;
+    NSError *error = nil;
+    // Use an autoreleasepool to release memory as soon as possible
+    @autoreleasepool {
+        writeResult = [NSJSONSerialization writeJSONObject:jsonObject
+                                                  toStream:stream
+                                                   options:0
+                                                     error:&error];
+    }
+
+    if (error) {
+        SENTRY_LOG_ERROR(@"Internal error while serializing JSON: %@", error);
+        return NO;
+    }
+    if (writeResult < 0) {
+        SENTRY_LOG_ERROR(@"Internal error while writing object to stream");
+        return NO;
+    }
+
+    return YES;
+}
+
++ (NSMutableDictionary *)buildEnvelopeSerialized:(SentryEnvelope *_Nonnull)envelope
+{
+    NSMutableDictionary *serializedData = [NSMutableDictionary dictionary];
+    if (nil != envelope.header.eventId) {
+        [serializedData setValue:[envelope.header.eventId sentryIdString] forKey:@"event_id"];
+    }
+
+    SentrySdkInfo *sdkInfo = envelope.header.sdkInfo;
+    if (nil != sdkInfo) {
+        [serializedData setValue:[sdkInfo serialize] forKey:@"sdk"];
+    }
+
+    SentryTraceContext *traceContext = envelope.header.traceContext;
+    if (traceContext != nil) {
+        [serializedData setValue:[traceContext serialize] forKey:@"trace"];
+    }
+
+    NSDate *sentAt = envelope.header.sentAt;
+    if (sentAt != nil) {
+        [serializedData setValue:sentry_toIso8601String(sentAt) forKey:@"sent_at"];
+    }
+    return serializedData;
+}
+
++ (NSData *)newlineData
+{
+    static NSData *newlineData = nil;
+    static dispatch_once_t onceToken;
+    dispatch_once(&onceToken, ^{ newlineData = [@"\n" dataUsingEncoding:NSUTF8StringEncoding]; });
+    return newlineData;
+}
+
++ (BOOL)writeEnvelope:(SentryEnvelope *)envelope toPath:(NSString *)path
```

Comment:
`m`: To be honest, I do not like that we are mixing a serialization utility class with actual file operations. IMO to have separation of concern, a serialize should ideally have two methods per type (serialize and deserialize X), mapping a type into data.
But as this is my opinion, I will not enforce this, but I would like to propose to move move most of the file management logic to e.g. `SentryFileManager`